### ISSUE TITLE
Extensible Sanitizer

### ISSUE THREAD

**Issue by Vandivier** [angular/angular#36650]
# 🚀 feature request

### Relevant Package
@angular/core

this line and file are of key interest:
https://github.com/angular/angular/blob/master/packages/core/src/sanitization/html_sanitizer.ts#L69

### Description
The ability to trust one or more keys or values within the sanitizer, without bypassing the whole sanitizer.

This can equally be described as an enhancement to the customizability of the sanitizer.

For one, this would allow me to preserve `id` attribute. I could also elect to allow `data` and/or `data-*` attributes with sanitizable values.

[Related SO 1](https://stackoverflow.com/questions/57400432/angular-how-to-retain-html-id-attribute-when-sanitized)
Related SO 2
this would close a bunch of github issues and resolve many SO questions.

### Describe the solution you'd like
I would like to be able to specify three things to allow keys and values to be sanitized in a SecurityContext.HTML.

This could be implemented as an options argument to DomSanitizer.sanitize, or it could be implemented as a custom SecurityContext, but I will use options object in my description below:

```
    trustedFirstParagraph$ = this.translateService
        .translate('creditwise.web.response.covid-faq.details.paragraphs.1')
        .pipe(map(s => this.domSanitizer.sanitize(SecurityContext.HTML, s, options)));
```

Here, options can be an object with any of three keys:
```
const options = {
  trustAttributeExpression: /some-regex-to-trust/,
  trustValueExpression: /some-value-to-trust/,
  trustKeyValueFunction: (el, key, value): boolean => {
      // arbitrary logic
  }
}
```

`trustAttributeExpression` and `trustValueExpression` must both match if both are specified. To implement an OR operation, and other complex algorithms, a developer can use `trustKeyValueFunction`.

### Describe alternatives you've considered
1. hiding data inside the class attribute and parsing it
2. using a span or div inside my element which is visually hidden but which I can access through my component dom ref and parse out the value
3. de-DRYing my code and having component-specific content across n components instead of being able to leverage a generic service

**Comment by gkgeorgiev** [comment]
Just about to request the same feature! My use-case is the callto:// URL prefix which I would like to whitelist.

**Comment by RichardMcSorley** [comment]
Agreed this would be a much need feature. I'm unable to use data attributes for my markdown to HTML project. So without this I'm blocked.

**Comment by Ruud-cb** [comment]
Please put this a bit higher on the to-do list, top-listed stackoverflow questions suggest using `bypassSecurityTrustHtml` as a solution! Thus leaving a lot of applications open for XSS attacks...

**Comment by cloakedninjas** [comment]
Came here looking for the same thing - but out of curiosity, what security concerns arise when `id` attributes are left in? Was this an oversight or a deliberate decision ?

**Comment by iwantwin** [comment]
> Came here looking for the same thing - but out of curiosity, what security concerns arise when `id` attributes are left in? Was this an oversight or a deliberate decision ?

When your code looks for a specific ID on the page and executes code based on what it found, and you have un-santized user-input higher up in the page, you can "steal" the element with the ID. Your code will look for the ID, and ID is supposed to be unique so the DOM will grab the first occurence, which is now controlled by the user.

As long as you are just changing some classes or something this is probably not vulnerable. But if the content or attributes of the associated element are used this could pose major threats. Or if a modal is shown to the user based on that ID? Could be used for showing a fake login screen on-site and sending the credentials to a culprits database.

**Comment by iwantwin** [comment]
I'm going to keep a close eye on this issue, as I think it would be great for Angular to have this functionality built in. Espescially for registered elements (customElements.define) would be great if we could "accept" those automatically.

In the meantime I've been using https://www.npmjs.com/package/sanitize-html which I wanted to share here. Highly customizable package that has been very reliable.

**Comment by Vandivier** [comment]
> When your code looks for a specific ID on the page and executes code based on what it found, and you have un-santized user-input higher up in the page, you can "steal" the element with the ID.

In my opinion this is neither a differentiator of the ID nor a security risk. It presumes the attacker is aware of the special ID and it's use case. The attacker might just as well be aware of an xpath or querySelector value that captures a special element. Indeed, ID seems to be de-risked in the sense that the attacker can only capture a single element using this pattern, where they might capture whole collections using a special querySelector.

In my view, element identification itself is not a security risk. It's the part you mention later; an attacker hijacking a user-controlled element, which is a very distinct and much more important concern for anyone familiar with web security. It's not a unique concern to the ID attribute and it's not solved by disabling ID sanitization.
